app/navigation/navigation_modal.js

```diff
@@ -24,16 +24,20 @@ export default class NavigationModal extends PureComponent {
     }
 
     componentWillReceiveProps(nextProps) {
-        if (this.props.show === nextProps.show) {
+        const nextRoutesLength = nextProps.children.props.navigationState.routes.length;
+        const currentRoutesLength = this.state.children ? this.state.children.props.navigationState.routes.length : 0;
+        const routesDidChange = nextRoutesLength !== currentRoutesLength;
```

Comment:
@jarredwitt @hmhealey Short of using lodash, is there a property of `navigationState.routes` that we could diff? What about `key`? Could we concatenate all the keys of the route stack and compare that?
```
const currentRouteKeys = this.state.children.props.navigationState.routes.map(r => r.key).join('');
const nextRouteKeys = nextProps.children.props.navigationState.routes.map(r => r.key).join('');
if (nextRouteKeys !== currentRouteKeys) {
```